### PR TITLE
TPC CDBInterface: pick specific file from CCDB with timestamp and met…

### PR DESCRIPTION
**Pull request by tklemenz**
…a data filter + newZSCalib helper function for TPC QC

This adds

- a function to get specific CalPad objects from the ccdb
- a new function to the helpers in TPC/qc to compare ZS calibration data

### CODE REVIEW

**Comment by tklemenz** [comment]
@wiechula Is the implementation in `BasicCCDBManager` and `CDBInterface` ok like this?

File: Detectors/TPC/base/include/TPCBase/CDBInterface.h [hunk 1]
```diff
@@ -123,8 +123,9 @@ class CDBInterface
   /// Return any CalPad object
   ///
   /// The function returns the CalPad object stored at the given path in the CCDB
-  /// \return CalPad object
-  const CalPad& getCalPad(const std::string_view path);
+  /// \return object
+  template <typename T>
+  T& getCalPad(const std::string_view path, long timestamp = -1, std::map<std::string, std::string> metaData = std::map<std::string, std::string>());
```
Comment (wiechula) [inline Detectors/TPC/base/include/TPCBase/CDBInterface.h]: What is the difference to `getSpecificObjectFromCDB`? looks like the exact same function. Perhaps one can remove this and simply make `getSpecificObjectFromCDB` public.

File: Detectors/TPC/qc/include/TPCQC/Helpers.h [hunk 2]
```diff
@@ -49,6 +51,8 @@ void setStyleHistogram2D(TH2& histo);
 /// Set nice style for vector of 2D histograms
 void setStyleHistogram2D(std::vector<TH2F>& histos);
 
+bool newZSCalib(const o2::tpc::CalDet<float>* refPedestal, const o2::tpc::CalDet<float>* refNoise, const o2::tpc::CalDet<float>* pedestal);
```
Comment (wiechula) [inline Detectors/TPC/qc/include/TPCQC/Helpers.h]:
```suggestion
bool newZSCalib(const o2::tpc::CalDet<float>& refPedestal, const o2::tpc::CalDet<float>& refNoise, const o2::tpc::CalDet<float>& pedestal);
```
you anyhow assume they are not nullpointers...

File: Detectors/TPC/qc/include/TPCQC/Helpers.h [hunk 3]
```diff
@@ -49,6 +51,8 @@ void setStyleHistogram2D(TH2& histo);
 /// Set nice style for vector of 2D histograms
 void setStyleHistogram2D(std::vector<TH2F>& histos);
```
Comment (wiechula) [inline Detectors/TPC/qc/include/TPCQC/Helpers.h:52]: please add some documentation what the function is supposed to be for.

**Review by wiechula — CHANGES_REQUESTED** [review]

File: Detectors/TPC/base/include/TPCBase/CDBInterface.h (hunk 1 above)
Comment (tklemenz) [reply Detectors/TPC/base/include/TPCBase/CDBInterface.h]: I agree, the `getCalPad` function can be removed. I also implemented the comments below in the latest pushed version.

File: Detectors/TPC/base/src/CDBInterface.cxx [hunk 4]
```diff
@@ -404,3 +398,7 @@ void CDBStorage::printObjectSummary(std::string_view name, CDBType const type, M
 
   LOGP(info, message);
 }
+
+//template CalPad& CDBInterface::getCalPad(const std::string_view path, long timestamp, std::map<std::string, std::string> metaData);
+//template std::vector<CalPad>& CDBInterface::getCalPad(const std::string_view path, long timestamp, std::map<std::string, std::string> metaData);
+//template std::unordered_map<std::string, o2::tpc::CalPad>& CDBInterface::getCalPad(const std::string_view path, long timestamp, std::map<std::string, std::string> metaData);
```
Comment (wiechula) [inline Detectors/TPC/base/src/CDBInterface.cxx]: can this be removed?

File: Detectors/TPC/qc/include/TPCQC/Helpers.h (hunk 3 above)
Comment (tklemenz) [reply Detectors/TPC/qc/include/TPCQC/Helpers.h:52]: Should I make a list in the docu of what it does to have it in doxygen directly? Isn't this comment sufficient and whoever wants to know what it does exactly just looks at the implementation?

**Review by wiechula — CHANGES_REQUESTED** [review]
Thanks for the update. Only minor things.

File: Detectors/TPC/base/src/CDBInterface.cxx (hunk 4 above)
Comment (tklemenz) [reply Detectors/TPC/base/src/CDBInterface.cxx]: oh, yes. This was an oversight on my side.

Comment (tklemenz) [reply Detectors/TPC/base/src/CDBInterface.cxx]: ok, done

**Review by wiechula — DISMISSED** [review]

**Comment by tklemenz** [comment]
The reason for the failing build in QualityControl will be fixed in [this PR](https://github.com/AliceO2Group/QualityControl/pull/746).

**Review by wiechula — DISMISSED** [review]

File: Detectors/TPC/base/include/TPCBase/CDBInterface.h [hunk 5]
```diff
@@ -236,6 +241,25 @@ inline T& CDBInterface::getObjectFromCDB(std::string_view path)
...
+inline T& CDBInterface::getSpecificObjectFromCDB(std::string_view path, long timestamp, std::map<std::string, std::string> metaData)
+{
+  static auto& cdb = o2::ccdb::BasicCCDBManager::instance();
+  auto* object = cdb.getSpecific<T>(path.data(), timestamp, metaData);
+  return *object;
+}
+
+template CalPad& CDBInterface::getSpecificObjectFromCDB(const std::string_view path, long timestamp, std::map<std::string, std::string> metaData);
+template std::vector<CalPad>& CDBInterface::getSpecificObjectFromCDB(const std::string_view path, long timestamp, std::map<std::string, std::string> metaData);
+template std::unordered_map<std::string, o2::tpc::CalPad>& CDBInterface::getSpecificObjectFromCDB(const std::string_view path, long timestamp, std::map<std::string, std::string> metaData);
```
Comment (shahor02) [inline Detectors/TPC/base/include/TPCBase/CDBInterface.h]: I guess the argument should be `const std::map<std::string, std::string>& metaData` ?

Comment (tklemenz) [reply Detectors/TPC/base/include/TPCBase/CDBInterface.h]: Good point, I changed it. The failing build seems to be unrelated as far as I can see, though.

**Comment by tklemenz** [comment]
The errors in `build/AliceO2/O2/o2/macOS` and `build/O2/fullCI` seem to be unrelated.

**Comment by shahor02** [comment]
@TimoWilken the fullCI complains on 
````
/sw/SOURCES/O2/6531-slc8_x86-64/0//Analysis/ALICE3/src/alice3-lutmaker.cxx:1:1: error: missing or malformed copyright notice
/sw/SOURCES/O2/6531-slc8_x86-64/0//Analysis/ALICE3/src/alice3-qa-singleparticle.cxx:1:1: error: missing or malformed copyright notice
````
but these files are not in this branch, they were added later. Looks like logs of different builds are mixed.

**Comment by davidrohr** [comment]
Hi @shahor02 : This happens when the build environment is reset / old or when a prerequisite was recompiled. The codechecker checks all files that were build during O2 compilation, not only those that are affected by the PR.

**Comment by shahor02** [comment]
@davidrohr ok, thanks

@tklemenz the problem failing macOS was already fixed, let's wait until it reruns.

**Review by wiechula — APPROVED** [review]
@shahor02 , it seems again the failures are unrelated. Do you have further suggestions?

**Comment by shahor02** [comment]
Actually, the macOS failure is from July 1, it did not rerun. OK, merging.

**Pull request merged**